sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedProducerAsyncClient.java

```diff
@@ -0,0 +1,326 @@
+    }
+
+    /**
+     * Enqueue iterable events without send options, will be sent to event hub by default send options.
+     *
+     * @param events Events will be stored in client temporarily and sent to event hub.
+     * @throws AmqpException if the size of {@code events} exceed the maximum size of a single batch.
+     * @return Gets a {@link Mono} that completes when eventData enqueued.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Void> enqueueEvents(Iterable<EventData> events) {
+        return enqueueEvents(events, DEFAULT_SEND_OPTIONS);
+    }
+
+    /**
+     * Enqueue iterable events with send options.
+     *
+     * @param events Events will be stored in client temporarily and sent to event hub.
+     * @param options A set of options used to configure the {@link EventDataBatch}.
+     * @throws AmqpException if the size of {@code events} exceed the maximum size of a single batch.
+     * @return Gets a {@link Mono} that completes when eventData enqueued.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Void> enqueueEvents(Iterable<EventData> events, SendOptions options) {
+        Objects.requireNonNull(events, "'eventData' cannot be null.");
+        CreateBatchOptions createBatchOptions = new CreateBatchOptions().setPartitionId(options.getPartitionId()).setPartitionKey(options.getPartitionKey());
+        AtomicReference<EventDataBatch> currentBatch = new AtomicReference<>();
+        ConcurrentLinkedQueue<EventData> failEvents = new ConcurrentLinkedQueue<>();
+
+        Flux<EventData> data = Flux.fromIterable(events);
+        return Mono.when(
+            data.flatMap(eventData -> {
+                EventDataBatch currentEventBatch = currentBatch.get();
+                if (Objects.isNull(currentEventBatch) || !currentEventBatch.tryAdd(eventData)) {
+                    if (!Objects.isNull(currentBatch.get())) {
+                        EventDataBatch fullBatch = currentBatch.getAndSet(null);
+                        addBatchToPartitionMap(options.getPartitionId(), fullBatch);
+                    }
+                    return producer.createBatch(createBatchOptions).map(eventDataBatch -> {
+                        currentBatch.set(eventDataBatch);
+                        if (!eventDataBatch.tryAdd(eventData)) {
+                            failEvents.add(eventData);
+                        }
+                        return eventDataBatch;
+                    });
+                }
+                return Mono.just(currentEventBatch);
+            }),
+            Mono.just(currentBatch.getAndSet(null)).map(eventDataBatch -> {
+                clientOptions.getSendFailedContext().accept(new SendBatchFailedContext(failEvents, options.getPartitionId(),
+                    new IllegalArgumentException(String.format(Locale.US, "Size of the payload exceeded maximum message size: %s kb", createBatchOptions.getMaximumSizeInBytes() / 1024))));
```

Comment:
What happens if some message failed to send? Do we throw exception, handler callback?

Reply:
In the publish part, I check the completion and send batch failed context.